src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/SpecialFileProviders/AbstractSpecialFileProvider.cs

```diff
@@ -0,0 +1,261 @@
+        private IProjectTree FindFile(string specialFileName)
+        {
+            IProjectTree rootNode = _projectTreeService.CurrentTree.Tree;
+
+            if (specialFileName == null)
+            {
+                return null;
+            }
+
+            IProjectTree specialFileNode;
+            // First, we look in the AppDesigner folder.
+            IProjectTree appDesignerFolder = rootNode.Children.FirstOrDefault(child => child.IsFolder && child.Flags.HasFlag(ProjectTreeFlags.Common.AppDesignerFolder));
+            if (appDesignerFolder != null && CreatedByDefaultUnderAppDesignerFolder)
+            {
+                specialFileNode = FindFileWithinNode(appDesignerFolder, specialFileName);
+                if (specialFileNode != null)
+                {
+                    return specialFileNode;
+                }
+            }
+
+            // Now try the root folder.
+            specialFileNode = FindFileWithinNode(rootNode, specialFileName);
+            if (specialFileNode != null)
+            {
+                return specialFileNode;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Find a file with the given filename within the given node.
+        /// </summary>
+        private IProjectTree FindFileWithinNode(IProjectTree parentNode, string fileName)
+        {
+            IProjectTree fileNode;
+            parentNode.TryFindImmediateChild(fileName, out fileNode);
+
+            if (fileNode != null)
+            {
+                // The user has created a folder with this name which means we don't have a special file.
+                if (fileNode.IsFolder)
+                {
+                    return null;
+                }
+
+                return fileNode;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Check to see if a given node is both part of a project and exists on disk. If asked to 
+        /// force sync, then either a file is included in the project or removed from the project.
+        /// </summary>
+        private async Task<bool> IsNodeInSyncWithDiskAsync(IProjectTree specialFileNode, bool forceSync, CancellationToken cancellationToken)
+        {
+            // If the file exists on disk but is not part of the project.
+            if (!specialFileNode.Flags.IsIncludedInProject())
+            {
+                if (forceSync)
+                {
+                    // Since the file already exists on disk, just include it in the project.
+                    var showAllFilesProvider = _projectTreeService.CurrentTree.TreeProvider as IShowAllFilesProjectTreeProvider;
+
+                    // Cannot include files if tree provider doesn't support it.
+                    if (showAllFilesProvider == null)
+                    {
+                        return false;
+                    }
+
+                    await showAllFilesProvider.IncludeItemsAsync(ImmutableHashSet.Create(specialFileNode));
```

Comment:
ie, why isn't this via IProjectItemProvider similar to what you did when you created a new item?